src/host-config/proxy.ts

```diff
@@ -0,0 +1,275 @@
+import { promises as fs } from 'fs';
+import path from 'path';
+import { isRight } from 'fp-ts/lib/Either';
+import Reporter from 'io-ts-reporters';
+
+import type { RedsocksConfig, HostProxyConfig } from './types';
+import { ProxyConfig } from './types';
+import { pathOnBoot, readFromBoot, writeToBoot } from '../lib/host-utils';
+import { unlinkAll, mkdirp } from '../lib/fs-utils';
+import { isENOENT } from '../lib/errors';
+import log from '../lib/supervisor-console';
+import * as dbus from '../lib/dbus';
+
+const proxyBasePath = pathOnBoot('system-proxy');
+const noProxyPath = path.join(proxyBasePath, 'no_proxy');
+const redsocksConfPath = path.join(proxyBasePath, 'redsocks.conf');
+
+export const disallowedProxyFields = ['local_ip', 'local_port'];
+export const validRedsocksProxyTypes = [
+	'socks4',
+	'socks5',
+	'http-connect',
+	'http-relay',
+];
+
+const isAllowedProxyField = (field: string): boolean =>
+	['type', 'ip', 'port', 'login', 'password'].includes(field);
+
+const isAuthField = (field: string): boolean =>
+	['login', 'password'].includes(field);
+
+// ? is a lazy operator, so only the contents up until the first `}(?=\s|$)` is matched.
+// (?=\s|$) indicates that `}` must be followed by a whitespace or end of file to match,
+// in case there are user fields with brackets such as login or password fields.
+const blockRegexFor = (blockLabel: string) =>
+	new RegExp(`${blockLabel}\\s?{([\\s\\S]+?)}(?=\\s|$)`);
```

Comment:
is `[\\s\\S]+` equivalent to `.+`? If not it might be better to make explicit what you are excluding, e.g. if you don't want numbers, you can do `[^0-9]`

Reply:
I think that `(?=\\s|$)` can be `\s*$` (i.e. any whitespace character or the end of the line)

Reply:
Overall the simpler version might be 
```suggestion
	new RegExp(`${blockLabel}\\s*{(.+?)}\\s*$)`);
```

But I may be missing something

Reply:
Regarding the suggestion of `\\s?` --> `\\s*`: I think it's better to keep as `\\s?` since it's more strict with its rule of "either one whitespace or nothing" as opposed to the rule of "0 or more whitespaces".

Regarding the suggestion of changing the capture group from `[\\s\\S]` --> `.`, I tried that before settling on `\\s\\S`, as `.` didn't pass for [this](https://github.com/balena-os/balena-supervisor/blob/91d090355559c3338287ac54d63a50cb10980264/test/unit/host-config.spec.ts#L130-L165) test case. I don't know the particulars of how JS applies different types of regex under the hood, but `\\s\\S` worked where `.` didn't, so something must be different.

Regarding suggestion `(?=\\s|$)` --> `\\s*$`, I also tried that initially but it also didn't pass the same test case as above, capturing `test = test` from the `dnstc` block following `redsocks`.

Feel free to play around with it yourself [here](regexr.com/82bmq).

Reply:
> Regarding the suggestion of \\s? --> \\s*: I think it's better to keep as \\s? since it's more strict with its rule of "either one whitespace or nothing" as opposed to the rule of "0 or more whitespaces".

I don't think it is necessarily to be more strict here. An extra space is easy to miss so we should be more permissive there

> Regarding the suggestion of changing the capture group from [\\s\\S] --> ., 

You are right, the dot does not match new lines without the [dotAll](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/dotAll) property, the code below works, but I think what you have is fine

```
new RegExp(`${blockLabel}\\s+{(.+?)}(?=\\s|$)`, 's');
```

> Regarding suggestion (?=\\s|$) --> \\s*$

You are right, that doesn't work. You are right is probably related to the lazy lookahead

Let's leave the code as is